neuralop/layers/coda_blocks.py

```diff
@@ -0,0 +1,417 @@
+        k = k.view(batch_size, self.n_heads, t, -1)
+        q = q.view(batch_size, self.n_heads, t, -1)
+        v = v.view(batch_size, self.n_heads, t, -1)
+
+        # attention mechanism
+        dprod = (torch.matmul(q, k.transpose(-1, -2)) /
+                 (np.sqrt(k.shape[-1]) * self.temperature))
+        dprod = F.softmax(dprod, dim=-1)
+
+        attention = torch.matmul(dprod, v)
+
+        # Reshape from (b, n, t, d * h * w) to (b, n, t, d, h, w, ...)
+        attention = attention.view(
+            attention.size(0),
+            attention.size(1),
+            attention.size(2),
+            d,
+            *tokens.shape[-self.n_dim:])
+        attention = torch.transpose(attention, 1, 2)
+        attention = attention.reshape(attention.size(0) * attention.size(1),
+                                   attention.size(2) * attention.size(3),
+                                   *attention.shape[-self.n_dim:])
+
+        return attention
+
+    def forward(self, x):
+        if self.permutation_eq:
+            return self._forward_equivariant(x)
+        else:
+            return self._forward_non_equivariant(x)
+
+    def _forward_equivariant(self, x):
+        """Use permutation equivariant mixer layer after the attention mechanism. 
+        Shares the same mixer layer for all the variables.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input tensor with shape (b, t * d, h, w, ...), where:
+            b is the batch size,
+            t is the number of tokens,
+            and d is the token codimension.
+        """
+        batch_size = x.shape[0]
+        output_shape = x.shape[-self.n_dim:]
+
+        assert x.shape[1] % self.token_codimension == 0, "Number of channels in x should be divisible by token_codimension"
+
+        # reshape from shape b (t*d) h w ... to (b*t) d h w ...
+        t = x.size(1) // self.token_codimension
+        tokens = x.view(
+            x.size(0) * t,
+            self.token_codimension,
+            *x.shape[-self.n_dim:])
+
+        # normalization and attention mechanism
+        tokens_norm = self.norm1(tokens)
+        attention = self.compute_attention(tokens_norm, batch_size)
+        if self.multi_head_proj is not None:
+            attention = self.multi_head_proj(attention)
+        attention = self.attention_normalizer(attention + tokens)
+        attention_normalized = self.norm2(attention)
+        output = self.mixer(attention_normalized, output_shape=output_shape)
+        output = self.mixer_out_normalizer(output) + attention
+
+        # reshape from shape (b*t) d h w... to b (t d) h w ...
+        t = output.size(0) // batch_size
+        output = output.view(
+            batch_size,
+            t * output.size(1),
+            *output.shape[-self.n_dim:])
+
+        return output
+
+    def _forward_non_equivariant(self, x):
+        """uses non-permuatation equivariant mixer
+        layer and normalizations.
+        Parameters
```

Comment:
There should be new line before and after the section title I believe for it to render properly. Would be good to check the doc when we add these. 

We should add this to the API doc as well.

Reply:
These won't be documented as they are private members. The main forward pass will be documented and we should add info on both forms of the forward pass (similar to preactivation vs. postactivation forward calls in FNOBlock)